### ISSUE TITLE
NPE at pro.bashsupport.shell.kc.getSourceRoots(kc.java:27)

### ISSUE THREAD

**Issue by kaleev** [BashSupport-Pro/bashsupport-pro#106]
Exception with the following stacktrace was reported:

```
Failed to execute task com.intellij.openapi.project.DumbServiceMergingTaskQueue$QueuedDumbModeTask@17bb7dda. null

java.lang.NullPointerException
	at pro.bashsupport.shell.kc.getSourceRoots(kc.java:27)
	at com.intellij.openapi.roots.impl.RootIndex.buildRootInfo(RootIndex.java:262)
	at com.intellij.openapi.roots.impl.RootIndex.<init>(RootIndex.java:72)
	at com.intellij.openapi.roots.impl.RootIndex.<init>(RootIndex.java:56)
	at com.intellij.openapi.roots.impl.DirectoryIndexImpl.getRootIndex(DirectoryIndexImpl.java:172)
	at com.intellij.openapi.roots.impl.DirectoryIndexImpl.getRootIndex(DirectoryIndexImpl.java:155)
	at com.intellij.openapi.roots.impl.DirectoryIndexImpl.getInfoForFile(DirectoryIndexImpl.java:182)
	at com.intellij.openapi.roots.impl.FileIndexBase.getInfoForFileOrDirectory(FileIndexBase.java:96)
	at com.intellij.openapi.roots.impl.ProjectFileIndexImpl.getInfoForFileOrDirectory(ProjectFileIndexImpl.java:27)
	at com.intellij.openapi.roots.impl.ProjectFileIndexImpl.isExcluded(ProjectFileIndexImpl.java:81)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods$shouldIndexFile$$inlined$runReadAction$1.compute(actions.kt:60)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:920)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods.shouldIndexFile(IndexableFilesIterationMethods.kt:71)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods.access$shouldIndexFile(IndexableFilesIterationMethods.kt:14)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods$iterateRoots$finalFileFilter$1.accept(IndexableFilesIterationMethods.kt:29)
	at com.intellij.openapi.vfs.VirtualFileFilter.lambda$and$0(VirtualFileFilter.java:50)
	at com.intellij.openapi.vfs.VfsUtilCore$1.visitFileEx(VfsUtilCore.java:277)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:293)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:325)
	at com.intellij.openapi.vfs.VfsUtilCore.iterateChildrenRecursively(VfsUtilCore.java:274)
	at com.intellij.openapi.vfs.VfsUtilCore.iterateChildrenRecursively(VfsUtilCore.java:267)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods.iterateRoots(IndexableFilesIterationMethods.kt:31)
	at com.intellij.util.indexing.roots.IndexableFilesIterationMethods.iterateRoots$default(IndexableFilesIterationMethods.kt:24)
	at com.intellij.util.indexing.roots.LibraryIndexableFilesIteratorImpl.iterateFiles(LibraryIndexableFilesIteratorImpl.kt:50)
	at com.intellij.util.indexing.UnindexedFilesUpdater.lambda$collectIndexableFilesConcurrently$15(UnindexedFilesUpdater.java:505)
	at com.intellij.openapi.roots.impl.PushedFilePropertiesUpdaterImpl.lambda$invokeConcurrentlyIfPossible$22(PushedFilePropertiesUpdaterImpl.java:389)
	at com.intellij.openapi.roots.impl.FilesScanExecutor.runOnAllThreads(FilesScanExecutor.java:68)
	at com.intellij.openapi.roots.impl.PushedFilePropertiesUpdaterImpl.invokeConcurrentlyIfPossible(PushedFilePropertiesUpdaterImpl.java:387)
	at com.intellij.util.indexing.UnindexedFilesUpdater.collectIndexableFilesConcurrently(UnindexedFilesUpdater.java:520)
	at com.intellij.util.indexing.UnindexedFilesUpdater.updateUnindexedFiles(UnindexedFilesUpdater.java:245)
	at com.intellij.util.indexing.UnindexedFilesUpdater.lambda$performScanningAndIndexing$18(UnindexedFilesUpdater.java:581)
	at com.intellij.util.gist.GistManagerImpl.runWithMergingDependentCacheInvalidations(GistManagerImpl.java:120)
	at com.intellij.util.indexing.UnindexedFilesUpdater.performScanningAndIndexing(UnindexedFilesUpdater.java:580)
	at com.intellij.util.indexing.UnindexedFilesUpdater.performInDumbMode(UnindexedFilesUpdater.java:571)
	at com.intellij.openapi.project.DumbServiceMergingTaskQueue$QueuedDumbModeTask.executeTask(DumbServiceMergingTaskQueue.java:211)
	at com.intellij.openapi.project.DumbServiceMergingTaskQueue$QueuedDumbModeTask.executeTask(DumbServiceMergingTaskQueue.java:195)
	at com.intellij.openapi.project.DumbServiceGuiTaskQueue.lambda$runSingleTask$1(DumbServiceGuiTaskQueue.java:64)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:189)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:176)
	at com.intellij.openapi.project.DumbServiceGuiTaskQueue.runSingleTask(DumbServiceGuiTaskQueue.java:62)
	at com.intellij.openapi.project.DumbServiceGuiTaskQueue.lambda$processTasksWithProgress$0(DumbServiceGuiTaskQueue.java:49)
	at com.intellij.util.io.storage.HeavyProcessLatch.performOperation(HeavyProcessLatch.java:79)
	at com.intellij.openapi.project.DumbServiceGuiTaskQueue.processTasksWithProgress(DumbServiceGuiTaskQueue.java:49)
	at com.intellij.openapi.project.DumbServiceImpl.lambda$runBackgroundProcess$13(DumbServiceImpl.java:616)
	at com.intellij.openapi.util.ShutDownTracker.executeWithStopperThread(ShutDownTracker.java:114)
	at com.intellij.openapi.project.DumbServiceImpl.runBackgroundProcess(DumbServiceImpl.java:612)
	at com.intellij.openapi.project.DumbServiceImpl$5.run(DumbServiceImpl.java:590)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$5(CoreProgressManager.java:493)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:244)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:189)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:176)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:244)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by jansorg** [comment]
@kaleev Thanks for reporting! Have you seen this error more than once? 
I'll make a fix for the next update, although I'm unable to reproduce it. According to the stacktrace it happened when the application was shut down.

**Comment by kaleev** [comment]
I haven't seen the error. We've just received several anonymous error reports from users. There is no additional comments, just a stacktrace. The reports are from CLion build 221.5080.93. Unfortunately there is no plugin version in the reports.

**Comment by jansorg** [comment]
Thanks! I didn't see that you're with JetBrains.
The NPE is caused by  `AdditionalLibraryRootsProvider` which use a project service. But during shutdown it's unavailable. I'm now checking `ShutDownTracker.isShutdownHookRunning`.
The fix will be part of 2.1.7, the next update of the stable version.
